### ISSUE TITLE
DataTable/TreeTable: cellEditor saves changes too late

### ISSUE THREAD

**Issue by VsevolodGolovanov** [primefaces/primefaces#7776]
**Describe the defect**
DataTable cellEditor saves changes when you click away from it. But if you click something like a commandButton, then the saving request gets queued after the button's, which is usually not what you want.

**Reproducer**
[primefaces-datatable-celledit-request-late.zip](https://github.com/primefaces/primefaces/files/7085911/primefaces-datatable-celledit-request-late.zip)

**Environment:**
 - PF Version: _6.2.30_, _10.0.0_
 - JSF + version: _ALL_
 - Affected browsers: _ALL_

**To Reproduce**
1. Go to http://localhost:8080/primefaces-test/
2. Click on the "Initial" value in the table.
3. Change it to "123". Don't press Enter.
4. Click on the button.
5. Note the stdout:
```
Acting on value: Initial
Changing value to 123
```
The value was changed after the button's action.

**Expected behavior**
Saving request happens before whatever was triggered by the click away. Maybe by also handling focusin or mousedown, or mouseup?

**Example XHTML**
```
<h:form id="form">
	<p:dataTable id="datatable" value="#{testView.list}" var="row" style="width: 100px" editable="true" editMode="cell">
		<p:column headerText="Column">
			<p:cellEditor>
				<f:facet name="input">
					<p:inputText value="#{row.value}" />
				</f:facet>
				<f:facet name="output">
					#{row.value}
				</f:facet>
			</p:cellEditor>
		</p:column>
	</p:dataTable>
	<p:commandButton id="btn" value="Act" action="#{testView.act}" process="@this" />
</h:form>

```

**Example Bean**
```java
@Data
@Named
@ViewScoped
public class TestView implements Serializable {

	private List<Bean> list = List.of(new Bean());

	public List<Bean> getList() {
		return list;
	}

	public void act() {
		System.out.println("Acting on value: " + list.get(0).getValue());
	}

	public static class Bean {

		private String value = "Initial";

		public String getValue() {
			return value;
		}

		public void setValue(String value) {
			System.out.println("Changing value to " + value);
			this.value = value;
		}

	}
    
}
```

**Comment by VsevolodGolovanov** [comment]
Btw the reproducer won't start with master-SNAPSHOT, same exception as here: https://github.com/primefaces/primefaces/issues/7604#issuecomment-882491021

**Comment by VsevolodGolovanov** [comment]
> focusin or mousedown, or mouseup

I know that if you use something like blockUI bound to ajax requests that this could swallow the click, and the button won't get it at all. But this is preferrable behavior.
Maybe make handled events configurable.

**Comment by melloware** [comment]
This has been reported a lot before.  To fix it locally you need AJAX on the input. Not sure how to fix it otherwise.
```xml
<p:inputText value="#{row.value}">
   <p:ajax process="@this" />
</p:inputText>
```

**Comment by VsevolodGolovanov** [comment]
In the real application the action invalidates the table's model, so when the cellEdit event arrives it ends with a system error.

It's fixable by handling some event that happens before the target click. mouseup is probably best, as focusin is not supported in older Firefox versions, and mousedown could be too late as well.

**Comment by melloware** [comment]
the only other option is on your CommandButton `onclick` is to use widget methods to close the celleditors first.

**Comment by melloware** [comment]
Please submit a PR with any proposed changes and we can look at it.  I know others have brought this issue up in the past.

**Comment by VsevolodGolovanov** [comment]
I just realized the real page already does the p:ajax workaround, but updating the datatable (i.e. row) so that the cellEditor immediately rerenders in output mode. Problem is, the cell is still considered to be in input mode by the datatable js, and the cellEdit fires on the next click regardless. Hmmm. What's the bug here? Should datatable rerender the cell in input mode, not output? It doesn't seem to keep track of this state currently, as in JSF state. Or rerender in output mode, but js should recognize this as the end of editing? This would circumvent cellEdit, but that's what you get for updating.

**Comment by melloware** [comment]
I am not sure there is a right answer but I think that is what this ticket was getting at: https://github.com/primefaces/primefaces/issues/3465

In the end i am not sure there is any 1 solution that make all use cases happy.

**Comment by melloware** [comment]
@VsevolodGolovanov is this similar to this issue: https://github.com/primefaces/primefaces/issues/7546 ??

**Comment by VsevolodGolovanov** [comment]
@melloware, the description is a bit hard to parse, but I think so.

**Comment by melloware** [comment]
OK I marked his as duplicate and kept yours open since yours includes a reproducer.

**Comment by VsevolodGolovanov** [comment]
@melloware, see MR #7842. Verified that this fixes the ordering.

**Comment by melloware** [comment]
Nice work!  I made one comment on your PR.

**Comment by VsevolodGolovanov** [comment]
In case someone needs a monkey patch. Suitable for PF versions of at least 6.2.30 to 11.0.0 (so far).

```
(function(orig) {
	PrimeFaces.widget.DataTable.prototype.bindEditEvents = function() {
		orig.call(this);

		var $this = this;

		if (this.cfg.editMode === 'cell') {
			$(document).off('click.datatable-cell-blur' + this.id);

			// save/cancel on mouseup to queue the event request before whatever was clicked reacts
			// noinspection DuplicatedCode
			$(document).off('mouseup.datatable-cell-blur' + this.id)
				.on('mouseup.datatable-cell-blur' + this.id, function(e) {
					// ignore if not editing
					if (!$this.currentCell)
						return;

					var currentCell = $($this.currentCell);
					var target = $(e.target);

					// ignore clicks inside edited cell
					if (currentCell.is(target) || currentCell.has(target).length)
						return;

					// ignore clicks inside input overlays like calendar popups etc
					var ignoredOverlay = '.ui-input-overlay, .ui-editor-popup, #keypad-div, .ui-colorpicker-container';
					// and menus - in case smth like menubutton is inside the table
					ignoredOverlay += ', .ui-datepicker-buttonpane, .ui-menuitem, .ui-menuitem-link';
					// and blockers
					ignoredOverlay += ', .ui-blockui, .blockUI';
					if (target.is(ignoredOverlay) || target.closest(ignoredOverlay).length)
						return;

					if ($.datepicker._datepickerShowing || $('.p-datepicker-panel:visible').length)
						return;

					if ($this.cfg.saveOnCellBlur)
						$this.saveCell($this.currentCell);
					else
						$this.doCellEditCancelRequest($this.currentCell);
				});
		}
	};
})(PrimeFaces.widget.DataTable.prototype.bindEditEvents);

(function(orig) {
	PrimeFaces.widget.TreeTable.prototype.bindEditEvents = function() {
		orig.call(this);

		var $this = this;

		if (this.cfg.editMode === 'cell') {
			$(document).off('click.treetable-cell-blur' + this.id);

			// save/cancel on mouseup to queue the event request before whatever was clicked reacts
			// noinspection DuplicatedCode
			$(document).off('mouseup.treetable-cell-blur' + this.id)
				.on('mouseup.treetable-cell-blur' + this.id, function(e) {
					// ignore if not editing
					if (!$this.currentCell)
						return;

					var currentCell = $($this.currentCell);
					var target = $(e.target);

					// ignore clicks inside edited cell
					if (currentCell.is(target) || currentCell.has(target).length)
						return;

					// ignore clicks inside input overlays like calendar popups etc
					var ignoredOverlay = '.ui-input-overlay, .ui-editor-popup, #keypad-div, .ui-colorpicker-container';
					// and menus - in case smth like menubutton is inside the table
					ignoredOverlay += ', .ui-datepicker-buttonpane, .ui-menuitem, .ui-menuitem-link';
					// and blockers
					ignoredOverlay += ', .ui-blockui, .blockUI';
					if (target.is(ignoredOverlay) || target.closest(ignoredOverlay).length)
						return;

					if ($.datepicker._datepickerShowing || $('.p-datepicker-panel:visible').length)
						return;

					if ($this.cfg.saveOnCellBlur)
						$this.saveCell($this.currentCell);
					else
						$this.doCellEditCancelRequest($this.currentCell);
				});
		}
	};
})(PrimeFaces.widget.TreeTable.prototype.bindEditEvents);
```

And/or if someone still needs the behavior where updating the table makes it forget the edited cell, so that no cellEdit/cellEditCancel event is fired.
```
(function(orig) {
	PrimeFaces.widget.DataTable.prototype.refresh = function(cfg) {
		orig.call(this, cfg);
		this.currentCell = null;
	};
})(PrimeFaces.widget.DataTable.prototype.refresh);

(function(orig) {
	PrimeFaces.widget.TreeTable.prototype.refresh = function(cfg) {
		orig.call(this, cfg);
		this.currentCell = null;
	};
})(PrimeFaces.widget.TreeTable.prototype.refresh);
```
